### PR TITLE
feature: typescript support, build optimizations and currency support

### PR DESCRIPTION
**Pull request by dreadera**
### **The Problem**
First of all I Apologize for the large chunk of changes. I tried to minimize it as much as possible.
This entire library has been stuck in a bit of a time machine. It's quite dated with failing tests, mismatched versioning and lack of proper development tools.

**The List of Problems:**
- Lack of TypeScript Support
- Failing Builds
- Failing Tests
- Webpack
- Chunky Bundle
- No Currency Support for the Currency Type Filter

### **The Goal**
- Update dependencies.
- Introduce new building tools.
- TypeScript support.
- Fix the buggy tests
- Add Multiple Currency / Locale Number Formatting

### **What has been done**
- Introduced Rollup as our default bundler. 
- Introduced TypeScript support.
- Introduced the useLocaleNumber hook.
- Introduced new Currency Filter Type Component.
- Introduced React Context Provider to act as a store.
- Existing Test Cases Updated.
- Existing Storybooks Updated.
- New Tests cases Added.
- Updated Storybooks.
- Updated Version Number.
- Dependency Clean up

### CODE REVIEW

**Comment by uptickmetachu** [comment]
My slight concern is we are going from one ball of complexity to another (webpack -> rollup). There seems to be equally a lot of configuration required WRT babel/eslint/rollup as there was before. 

I do appreciate the better build times and bundle size. I wonder if there is a way to simplify the build? Also have we tested this with `yarn link` against workforce to make sure the built version works with workforce (minus the janky stuff in workforce webpack).

File: package.json [hunk 1]
```diff
@@ -80,11 +77,19 @@
...
+    "rollup-plugin-peer-deps-external": "^2.2.4",
+    "rollup-plugin-terser": "^7.0.2",
+    "rollup-plugin-typescript2": "^0.30.0",
+    "rollup-plugin-visualizer": "^5.5.0",
+    "sass": "^1.33.0",
+    "sass-loader": "^11.1.1",
+    "style-loader": "^2.0.0",
+    "ts-jest": "^26.5.6",
+    "typescript": "^4.2.4",
+    "typescript-eslint": "^0.0.1-alpha.0"
```
Comment (uptickmetachu) [inline package.json:92]: o,o is this the correct plugin? 
https://www.npmjs.com/package/@typescript-eslint/eslint-plugin
Perhaps we mean this one?

File: rollup.config.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+import packageJson from './package.json'
+
+export default {
+  input: 'src/index.ts',
+  // All of your library files will be named exports from here.
+  output: [
+    // Common JS Support.
+    {
+      file: packageJson.main,
```
Comment (uptickmetachu) [inline rollup.config.js:20]: oh neat! I like this importing the package.json and referencing the file for linking the build tools to the packaging. 👍🏼

File: rollup.config.js [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+          'node_modules',
+          'dist',
+          'scripts',
+          'jest',
+          '**/*.test.ts',
+          '**/*.test.tsx'],
+      },
+    }),
+
+    // Do Babel transpilation.
```
Comment (uptickmetachu) [inline rollup.config.js:61]: :( Somehow I assumed we could get rid of babel if we adopted TS.

**Review by uptickmetachu — COMMENTED** [review]
Same as my comments above, otherwise looking good!

Regarding the actual frontend changes can you add a comment to sections that reviewers should look at? It has become hard to navigate due to all the format changes.

File: package.json (hunk 1 above)
Comment (dreadera) [reply package.json:92]: oops, you're right. I will rectify this :)

File: rollup.config.js (hunk 3 above)
Comment (dreadera) [reply rollup.config.js:61]: I fully agree that going from webpack -> rollup doesn't change things in terms of configuration and does add a bit of complexity if you don't know rollup but i chose it simply for the sake of having a large amount of community maintained plugins that work out of the box without much configuration. while the actually rollup config itself does require you incorporate the plugins, the plugins themselves don't need much configuration. 

if it makes things easier, we can try esbuilding everything without having rollup :) the configuration is also minimal for that :))

**Comment by dreadera** [comment]
I fully agree that going from webpack -> rollup doesn't change things in terms of configuration and does add a bit of complexity if you don't know rollup but i chose it simply for the sake of having a large amount of community maintained plugins that work out of the box without much configuration. while the actually rollup config itself does require you incorporate the plugins, the plugins themselves don't need much configuration.

if it makes things easier, we can try esbuilding everything without having rollup :) the configuration is also minimal for that :))

**Comment by dreadera** [comment]
In terms of testing with workforce, I haven't fully tried every place where this library is used but I did yarn link and have a go on a demo version I setup, which worked as intended :)

File: src/filters/types/__tests__/Date.test.js [hunk 4]
```diff
@@ -1,6 +1,6 @@
 import React from 'react'
 import { shallow } from 'enzyme'
-import { snapshotTest } from 'utils/tests'
+import { snapshotTest } from '../../../../utils/tests'
```
Comment (adamdickinson) [inline src/filters/types/__tests__/Date.test.js:3]: Out of curiosity - what benefit do we get from switching away from the aliased paths?

File: src/filters/types/__tests__/__snapshots__/Currency.test.js.snap [hunk 5]
```diff
@@ -1,25 +1,32 @@
 // Jest Snapshot v1, https://goo.gl/fbAQLP
 
+exports[`Currency Snapshots enders default 1`] = `
```
Comment (adamdickinson) [inline src/filters/types/__tests__/__snapshots__/Currency.test.js.snap:3]: Typo?

File: src/filters/types/__tests__/__snapshots__/currency.test.tsx.snap [hunk 6]
```diff
@@ -0,0 +1,16 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`Currency Snapshots enders default 1`] = `
```
Comment (adamdickinson) [inline src/filters/types/__tests__/__snapshots__/currency.test.tsx.snap:3]: Typo here too?

File: src/hooks/use-locale-number.test.tsx [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+  const { format } = useLocaleNumber({ ...parameters, currency: true })
+  return <div>{format(value)}</div>
+}
+
+const NumberFormatter: React.FC<Formatter> = ({ parameters, value }: Formatter) => {
+  const { format } = useLocaleNumber({ ...parameters })
+  return <div>{format(value)}</div>
+}
+
+describe('Convert Number To Locale', () => {
```
Comment (adamdickinson) [inline src/hooks/use-locale-number.test.tsx:22]: Question on consistency - we use lowercase in our other test files, but Title Case throughout this one. I think the convention is the former. Could we update to lowercase to stay with convention?

```suggestion
describe('convert number to locale', () => {
```

File: src/hooks/use-locale-number.ts [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+  const context = useContext(ObjectListContext)
+
+  const configuration = {
+    currency: false,
+
+    // add override from context if provided.
+    ...context,
+
+    // add override if params are provided.
+    ...params,
```
Comment (adamdickinson) [inline src/hooks/use-locale-number.ts:15]: I think you've done a good enough job making the code explain this for you :)
```suggestion
    ...context,
    ...params,
```

File: src/hooks/use-locale-number.ts [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+      return new Intl.NumberFormat(locale, options).format(0)
+    }
+
+    return new Intl.NumberFormat(locale, options).format(value)
+  }
+
+  return {
+    format,
+    configuration,
+  }
```
Comment (adamdickinson) [inline src/hooks/use-locale-number.ts:41]: Could we default to 0 to simplify this?

```suggestion
  function format(value = 0) {
    return new Intl.NumberFormat(locale, options).format(value)
  }

  return {
    format,
    configuration,
  }
```

File: src/utils/translate.ts [hunk 10]
```diff
@@ -0,0 +1,12 @@
...
+    String.prototype.split
+      .call(path, regexp)
+      .filter(Boolean)
+      .reduce((res, key) => (res !== null && res !== undefined ? res[key] : res))
+
+  const result = travel(/[,[\]]+?/) || travel(/[,[\].]+?/)
+  const payload = result === undefined
+
+  return payload
+}
```
Comment (adamdickinson) [inline src/utils/translate.ts:12]: What... does this function do?

**Review by adamdickinson — CHANGES_REQUESTED** [review]
Relatively quick pass from me. Awesome attention to detail in many parts - loving what you've done! I've noted a few things, queried a few parts and suggested some things along the way.

I will flag that I'm concerned we won't be able to give this a thorough and proper review due to its size, but I'll do my best :)

Also apologies for the short/sharp/inconsiderate comments! Loving your work, just short on time this morning 😬.

File: src/utils/translate.ts (hunk 10 above)
Comment (dreadera) [reply src/utils/translate.ts:12]: Ah, I forgot to remove this. I was planning to write a locale function to get strings from objects. like lodash/get. however, i moved away from this. I just forgot to remove the function.

File: src/hooks/use-locale-number.ts (hunk 9 above)
Comment (dreadera) [reply src/hooks/use-locale-number.ts:41]: Oh for sure, this makes more sense! :)

File: package.json [hunk 11]
```diff
@@ -1,77 +1,74 @@
 {
   "name": "react-object-list",
-  "version": "0.2.10",
+  "version": "0.3.00",
```
[the remark this answers is not in the document]
Comment (dreadera) [reply package.json:3]: I only changed the middle value to represent the minor update in accordance with SemVer to represent the new currency feature.

File: src/ObjectList.demo.stories.js [hunk 12]
```diff
@@ -1,25 +1,42 @@
...
-import {FontAwesome} from './icons'
+import { TextContainsFilter, SearchFilter } from './filters'
+import { FontAwesome } from './icons'
 
 import ObjectList from '.'
 import Table from './data-renderers/Table'
 import List from './data-renderers/List'
 
-const mockData = require('./demo.data.json')
+const mockData = require('../samples/people.json')
```
[the remark this answers is not in the document]
Comment (dreadera) [reply src/ObjectList.demo.stories.js:11]: We should. Faker is really nice for dev / showcasing :)

File: .eslintrc [hunk 13]
```diff
@@ -1,41 +1,84 @@
 {
-  "plugins": [
-    "babel"
-  ],
-  "parser": "babel-eslint",
+  "plugins": ["babel", "react", "react-hooks", "@typescript-eslint"],
```
[the remark this answers is not in the document]
Comment (dreadera) [reply .eslintrc:2]: I was thinking the same thing. I think we should all have a standardised version which should be version controlled and then be available via npm :)

File: src/filters/types/Currency.tsx [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+            readOnly
+          />
+      }
+    </div>
+  )
+}
+
+Currency.defaultProps = {
+  currencySymbol: 'AUD',
+}
```
Comment (jarekwg) [inline src/filters/types/Currency.tsx:50]: This doesn't look right -- "currencySymbol" suggests '$', not 'AUD'..
The way I'd see this working is the widget reads in locale info from a higher context, and then applies the appropriate

**Review by jarekwg — CHANGES_REQUESTED** [review]
I'm not strong on reviewing frontend code, so just going to provide cursory comments.
- The size of this PR is very hard to review properly. If it's relatively straightforward to cut a couple more branches from master, cherry pick individual changes into them, and open up a couple of smaller PRs. This is especially important when it comes to codestyle changes vs business logic changes. At the very least, the currency changes need to be moved out of here and done separately.
- My vote would be to use esbuild over rollup here too, just so that we're not spreading ourselves thing on bundler prerequisite knowledge..

File: src/filters/types/Currency.tsx (hunk 14 above)
Comment (dreadera) [reply src/filters/types/Currency.tsx:50]: The native Intl.NumberFormat function takes in the standard currency symbols.

This works a lot better as it's built into JS and means we only have to provide the following. Locale, Currency Format and Decimals to the context provider upon invocation.  

Function: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl/NumberFormat/NumberFormat
List: https://www.currency-iso.org/en/home/tables/table-a1.html

**Comment by dreadera** [comment]
I 100% agree that this was should have been broken down into multiple PRs. I do apologise for that.

File: src/filters/types/Currency.tsx (hunk 14 above)
Comment (jarekwg) [reply src/filters/types/Currency.tsx:50]: gotcha; can we rename it just to `currency` then?

**Comment by jarekwg** [comment]
Ok, if not planning to split this up, we'll want to publish this with at least a minor version release bump (ie 0.3.0). Then will probs be best to test by plugging it directly into workforce and clicking through to see how it all works. I suspect there'll need to be followup patch versions needed for the stuff we haven't picked up in this review.

**Comment by scaredcat** [comment]
The changes here won't quite work. This package is set up kind of like a PWA where
* `src/index` is the skeleton app. Placeholders for the filers (top section) and renderer (table/list)
* `src/filters/index` is the filters to display at the top of the app
* `src/data-renderers/index` is the content of the app (also imported directly into the skeleton app -- bad practice, but programmers are lazy. 99% of the time we will use Table renderer)
* `src/icons` -- us trying to be fancy with too much customization. We should just use one icon set and stick with that.
* `src/types` aka cells -- fine grained control of the type of data rendered within a table

In your changes `rollup.config.js` only references `src/index` which also includes the default data renderer and probably some cells. Its tree-shaking functionality is quite good so it will get rid of things that are not exported/referenced.

Please either
1. Export all the things from `src/index` to make sure it comes with all the necessary parts (preferred option)
2. Update rollup to continue delivering this package as a Frankenstein for self-assembly at the other end

**Comment by dreadera** [comment]
wow, thanks for all that. it's nice to get some feedback on someone who worked on it :)
I'll improve this PR with the notes you have given. Thank you @scaredcat

**Pull request closed without merging**